cryptonight/src/hash_v2.rs

```diff
@@ -0,0 +1,456 @@
+    chunk3[0..8].copy_from_slice(&(chunk2_old[0].wrapping_add(a0[0])).to_le_bytes());
+    chunk3[8..16].copy_from_slice(&(chunk2_old[1].wrapping_add(a0[1])).to_le_bytes());
+
+    let b0 = block_to_u64le(subarray(b, 0));
+    let chunk2 = &mut long_state[chunk2_start..chunk2_start + 16];
+    chunk2[0..8].copy_from_slice(&(chunk1_old[0].wrapping_add(b0[0])).to_le_bytes());
+    chunk2[8..16].copy_from_slice(&(chunk1_old[1].wrapping_add(b0[1])).to_le_bytes());
+
+    if variant == Variant::R {
+        let mut out_copy = block_to_u64le(c1);
+
+        chunk1_old[0] ^= chunk2_old[0];
+        chunk1_old[1] ^= chunk2_old[1];
+        out_copy[0] ^= chunk3_old[0];
+        out_copy[1] ^= chunk3_old[1];
+        out_copy[0] ^= chunk1_old[0];
+        out_copy[1] ^= chunk1_old[1];
+
+        c1[0..8].copy_from_slice(&out_copy[0].to_le_bytes());
+        c1[8..16].copy_from_slice(&out_copy[1].to_le_bytes());
+    }
+}
+
+#[expect(
+    clippy::cast_sign_loss,
+    clippy::cast_precision_loss,
+    clippy::cast_possible_truncation
+)]
+pub(crate) fn variant2_integer_math_sqrt(sqrt_input: u64) -> u64 {
+    // Get an approximation using floating point math
+    let mut sqrt_result =
+        ((sqrt_input as f64 + 18_446_744_073_709_552_000.0).sqrt() * 2.0 - 8589934592.0) as u64;
+
+    // Fixup the edge cases to get the exact integer result. For more information,
+    // see: https://github.com/monero-project/monero/blob/v0.18.3.3/src/crypto/variant2_int_sqrt.h#L65-L152
+    let sqrt_div2 = sqrt_result >> 1;
+    let lsb = sqrt_result & 1;
+    let r2 = sqrt_div2
+        .wrapping_mul(sqrt_div2 + lsb)
+        .wrapping_add(sqrt_result << 32);
+
+    if r2.wrapping_add(lsb) > sqrt_input {
+        sqrt_result = sqrt_result.wrapping_sub(1);
+    }
+    if r2.wrapping_add(1 << 32) < sqrt_input.wrapping_sub(sqrt_div2) {
+        // Not sure that this is possible. I tried writing a test program
+        // to search subsets of u64 for a value that can trigger this
+        // branch, but couldn't find anything. The Go implementation came
+        // to the same conclusion:
+        // https://github.com/Equim-chan/cryptonight/blob/v0.3.0/arith_ref.go#L39-L45
+        sqrt_result = sqrt_result.wrapping_add(1);
+    }
+
+    sqrt_result
+}
+
+/// Original C code:
+/// <https://github.com/monero-project/monero/blob/v0.18.3.4/src/crypto/slow-hash.c#L277-L283>
+pub(crate) fn variant2_integer_math(
+    c2: &mut [u8; 8],
+    c1: &[u8; AES_BLOCK_SIZE],
+    division_result: &mut u64,
+    sqrt_result: &mut u64,
+    variant: Variant,
+) {
+    const U32_MASK: u64 = u32::MAX as u64;
+
+    if variant != Variant::V2 {
+        return;
+    }
+
+    let tmpx = *division_result ^ (*sqrt_result << 32);
+    *c2 = (u64::from_le_bytes(*c2) ^ tmpx).to_le_bytes();
+
+    let c1_64 = block_to_u64le(c1);
+    let mut divisor = c1_64[0];
+    let dividend = c1_64[1];
+
+    divisor = ((divisor + ((*sqrt_result << 1) & U32_MASK)) | 0x80000001) & U32_MASK;
```

Comment:
I got an overflow here in debug mode

Reply:
Yeah, that should have been using `wrapping_add`. All this hash code was designed with intentional integer overflow and I was probably overcautious of not adding `wrapping_add` in places I didn't think overflow could be triggered. If you saved the hash input, please share it. I added a contrived unit test directly to `variant2_integer_math` that would trigger the previous code.

Reply:
Sadly I don't, the panic didn't cause it to print the hash input